### ISSUE TITLE
[compiler-rt][tysan][libc++] False positives with std::array element access

### ISSUE THREAD

**Issue by sfc-gh-oisler** [llvm/llvm-project#208650]
# `std::array`
#### Code: 
```.cpp
#include <array>

std::array<int, 3> a;

int main() {
    a[0] = 1;
}
```
#### Logs:
```
==1==ERROR: TypeSanitizer: type-aliasing-violation on address 0x61c8a6194ee4 (pc 0x61c8a583af84 bp 0x7ffcc1fa7320 sp 0x7ffcc1fa72b0 tid 1)
WRITE of size 4 at 0x61c8a6194ee4 with type int accesses an existing object of type std::__1::array<int, 3ul>
    #0 0x61c8a583af83  (/app/output.s+0x35f83)
```
#### Explanation:
libc++ `std::array` is likely firing due to wrapper vs raw element storage

**Comment by firedog1234** [comment]
@zwuis

Hello, I looked at this a bit, heres what I think.

The root cause seems to be that array struct fields fall back to "omnipotent char" in TBAA metadata under `-fsanitize=type` (the collapse-to-element-type rule is gated behind a flag which tysan never sets) sooo

You could do either:

1. enable that collapse for tysan. Fixes scalar/pointer arrays, but arrays of struct-typed elements can still have a false positive

   or

2. add a repeated element descriptor to tysan's runtime so offset checks use modular arithmetic instead of exact match so it works generally.

I think solution 2 would be more robust

**Comment by chrchr-github** [comment]
Reduced:
```c++
struct S {
    int e[1];
};

S s;

int main() {
    s.e[0] = 1;
}
```
https://godbolt.org/z/P8cWs74c8

**Comment by AmarD98** [comment]
I'm taking a look at this thanks for the writeup @firedog1234

**Comment by TheLastRar** [comment]
There is also https://github.com/llvm/llvm-project/issues/122954 which looks identical to the reduced case.
